encoding/dot/decode.go

```diff
@@ -0,0 +1,223 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dot
+
+import (
+	"fmt"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/encoding/dot/dotparser/ast"
+)
+
+// Builder is a graph that can have user-defined nodes and edges added.
+type Builder interface {
+	graph.Graph
+	graph.Builder
+	// NewNode adds a new node with a unique node ID to the graph.
+	NewNode() graph.Node
+	// NewEdge adds a new edge from the source to the destination node to the
+	// graph, or returns the existing edge if already present.
+	NewEdge(from, to graph.Node) graph.Edge
+}
+
+// UnmarshalerAttr is the interface implemented by objects that can unmarshal a
+// DOT attribute description of themselves.
+type UnmarshalerAttr interface {
+	// UnmarshalDOTAttr decodes a single DOT attribute.
+	UnmarshalDOTAttr(attr Attribute) error
+}
+
+// Copy copies the nodes and edges from the Graphviz AST source graph to the
+// destination graph. Edge direction is maintained if present.
+func Copy(dst Builder, src *ast.Graph) (err error) {
+	defer func() {
+		switch e := recover().(type) {
+		case nil:
+		case error:
+			err = e
+		default:
+			panic(e)
+		}
+	}()
+	gen := &generator{
+		directed: src.Directed,
+		ids:      make(map[string]graph.Node),
+	}
+	for _, stmt := range src.Stmts {
+		gen.addStmt(dst, stmt)
+	}
+	return err
+}
+
+// A generator keeps track of the information required for generating a gonum
+// graph from a dot AST graph.
+type generator struct {
+	// Directed graph.
+	directed bool
+	// Map from dot AST node ID to gonum node.
+	ids map[string]graph.Node
+	// Nodes processed within the context of a subgraph, that is to be used as a
+	// vertex of an edge.
+	subNodes []graph.Node
+	// Stack of start indices into the subgraph node slice. The top element
+	// corresponds to the start index of the active (or inner-most) subgraph.
+	subStart []int
+}
+
+// node returns the gonum node corresponding to the given dot AST node ID,
+// generating a new such node if none exist.
+func (gen *generator) node(dst Builder, id string) graph.Node {
+	if n, ok := gen.ids[id]; ok {
+		return n
+	}
+	n := dst.NewNode()
+	gen.ids[id] = n
+	// Check if within the context of a subgraph, that is to be used as a vertex
+	// of an edge.
+	if gen.isInSubgraph() {
+		// Append node processed within the context of a subgraph, that is to be
+		// used as a vertex of an edge
+		gen.appendSubgraphNode(n)
+	}
+	return n
+}
+
+// addStmt adds the given statement to the graph.
+func (gen *generator) addStmt(dst Builder, stmt ast.Stmt) {
+	switch stmt := stmt.(type) {
+	case *ast.NodeStmt:
+		n := gen.node(dst, stmt.Node.ID)
+		if n, ok := n.(UnmarshalerAttr); ok {
+			for _, attr := range stmt.Attrs {
+				a := Attribute{
+					Key:   attr.Key,
+					Value: attr.Val,
+				}
+				if err := n.UnmarshalDOTAttr(a); err != nil {
```

Comment:
I also felt this before, and wrote my code as:

```go
err := foo()
if err != nil {}
```

before, rather than.

```go
if err := foo() err != nil {}
```

However, as with may things in Go, I tried to learn from the standard library, and while initially I did not feel it was right, I learnt to appreciate it. Same with the placement of braces. Coming from C, I always placed them on their own line. Now having been using Go with gofmt for such a long time, I even started writing C code as

```c
int f() {
}
```

rather than

```c
int f()
{
}
```

The reason for placing return values in the *init* statement of if-statements, is to limit their scope.

Taking a quick look at the Go standard library, it seems that "if err := foo(); err != nil {}" is more common than "err := foo()"

```bash
[~/go/src]$ grep -R -P "\tif err :=" . | wc -l
1323
[~/go/src]$ grep -R -P "\terr :=" . | wc -l
524
```

That being said, I don't care too much, and if the prefered style of gonum is to keep the condition on a separate line, I'd can easily update the PR. Just wanted to give some background.